### ISSUE TITLE
Additional supported expression function datediff 

### ISSUE THREAD

**Issue by tnavda** [vega/vega#827]
In reference to the question raised around #822.  The problem with my solution is the labels are bound to what are statically entered.  So in the event that you wanted to zoom in on the data, you are still left with only the statically coded label values.  Is there a better, supported, way to do this without the addition of a new expression function?

signature:  datediff(datum.value, [hardcoded date], [return fmt i.e. s, m, h, etc])

Then use the d3.timeFormat string in the "format" property to specify our display of the time difference.

Which leads me in to another possible feature request; if by default our format string is "%H:%M" with our current domain, but when we zoom in we want "%H:%M:%S" to become the new format.  I don't know if conditional formats like this are maybe this is already possible with a signal.

Below are two examples Mike Bostock did around the topic in D3
https://bl.ocks.org/mbostock/3048166
https://bl.ocks.org/mbostock/c2b2a129696be78830d8

Does Vega support javascript functions in this scenario, versus a Vega expression, or does that then defeat the overall goal of Vega?

**Comment by jheer** [comment]
How about doing the following within an expression?
`utcFormat(+date2 - +date1, '%H:%M')`

This simply ensures that (1) dates are mapped to timestamps, (2) computes the difference, and (3) formats the resulting time offset using UTC time. The reason UTC (universal) time is needed is to ensure that a timestamp value of `0` maps to midnight (and thus the hours, minutes, and seconds components are all zero), which is not guaranteed depending on your local timezone.

For example, the following evaluates to `05:12` as one would like:
`utcFormat(+datetime(2017, 4, 1, 5, 12, 0) - +datetime(2017, 4, 1, 0, 0, 0), '%H:%M')`

Alternatively, you could compute the date differences only, and then visualize them using a scale of type `utc` to ensure appropriate time labels are created along your axis. If you leave the axis `format` parameter unspecified, you will get labels that automatically adjust format as you zoom in and out.

**Comment by tnavda** [comment]
There seems to be some issue with the starting X-axis value of 00:00 when I try it.  Below are two screenshots, the first with the lookup and the second with the signal.  You can see when using the signal the 00:00 doesn't start at the beginning of the X value, even though the date I am subtracting in my expression is the first value.  I've tried setting the domain to specific start and end times as well but the 00:00 is still shifted off the X value.

`      "domain": {"data": "table", "field": "x"}, `
`      "domain": [1495053866000, 1495057676000], `

`"text": {"signal": "utcFormat(datum.value - +1495053866000, '%H:%M')"} `

![without_signal](https://cloud.githubusercontent.com/assets/28634090/26422264/1ab42582-4097-11e7-91fd-46256e9edf18.png)

![with_signal](https://cloud.githubusercontent.com/assets/28634090/26422269/21281a2c-4097-11e7-9a7b-6a12cf49d733.png)

**Comment by jheer** [comment]
OK, I'd need to see the actual specifications in order to debug.

**Comment by tnavda** [comment]
Thanks for the help, attaching both versions.  Signal first, and lookup second

 var vegaJSONSpec = {
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "width": 778,
  "height": 389,
  "padding": 50,
  "signals": [
    {
      "name": "interpolate",
      "value": "step",
      }     
    ],
      
  "data": [
    {
      "name": "table",
      "format": {"type":"json", "parse":{"x":"number"}},
      "values": [     
    
{x: 1495053866000 , y:5},
{x: 1495053871000 , y:8},
{x: 1495053876000 , y:10},
{x: 1495053881000 , y:13},
{x: 1495053886000 , y:15},
{x: 1495053891000 , y:17},
{x: 1495053896000 , y:20},
{x: 1495053901000 , y:22},
{x: 1495053906000 , y:25},
{x: 1495053911000 , y:27},
{x: 1495053916000 , y:30},
{x: 1495053921000 , y:32},
{x: 1495053926000 , y:35},
{x: 1495053931000 , y:37},
{x: 1495053936000 , y:40},
{x: 1495053941000 , y:42},
{x: 1495053946000 , y:45},
{x: 1495053951000 , y:47},
{x: 1495053956000 , y:50},
{x: 1495057561000 , y:49},
{x: 1495057566000 , y:46},
{x: 1495057571000 , y:45},
{x: 1495057576000 , y:42},
{x: 1495057581000 , y:39},
{x: 1495057586000 , y:38},
{x: 1495057591000 , y:33},
{x: 1495057596000 , y:29},
{x: 1495057601000 , y:27},
{x: 1495057606000 , y:24},
{x: 1495057611000 , y:19},
{x: 1495057616000 , y:19},
{x: 1495057621000 , y:18},
{x: 1495057626000 , y:16},
{x: 1495057631000 , y:14},
{x: 1495057636000 , y:12},
{x: 1495057641000 , y:9},
{x: 1495057646000 , y:6},
{x: 1495057651000 , y:5},
{x: 1495057656000 , y:3},
{x: 1495057661000 , y:2},
{x: 1495057666000 , y:1},
{x: 1495057671000 , y:1},
{x: 1495057676000 , y:0},

                  
      ]
    }
  ],

  "scales": [
    {
      "name": "x",
      "type": "utc",
      "range": "width",
      "domain": {"data": "table", "field": "x"},
      //"domain": [1495053866000, 1495057676000],
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "nice": true,
      "zero": true,
      "domain": {"data": "table", "field": "y"}      
    },
    {
      "name": "color",
      "type": "ordinal",
      "range": "category",
      "domain": {"data": "table", "field": "c"}
    }   
  ],

  "axes": [
    {
      "orient": "bottom", 
      "scale": "x",    
      "format": "%H:%M",
      "grid": true,
      "titlePadding": 10,
      "encode": {
        "title": {
          "update": {
            "fontSize": {"value": 14},
            "font": {"value": "Verdana, Arial"},
          }
        },
        "labels": {
          "update": {
            "fontSize": {"value": 10},
            "font": {"value": "Verdana, Arial"},
            "text": {"signal": "utcFormat(+datum.value - +1495053866000, '%H:%M')"}
            }
          }          
      }
    },
    {"orient": "left", "scale": "y", "grid": true, "title": "Count","titlePadding": 10,
      "encode": {
        "title": {
          "update": {
            "fontSize": {"value": 14},
            "font": {"value": "Verdana, Arial"},
          
          }
        },
        "labels": {
          "update": {
            "fontSize": {"value": 10},
            "font": {"value": "Verdana, Arial"},
            }
          }          
      }
  }
  ],  
  "marks": [  
    {
      "type": "group",
      "from": {
        "facet": {
          "name": "series",
          "data": "table",
          "groupby": "c"
        }
      },
      "marks": [              
        {
          "type": "line",
          "from": {"data": "series"},
          "encode": {
            "enter": {
              "x": {"scale": "x", "field": "x"},
              "y": {"scale": "y", "field": "y"},
              "strokeWidth": {"value": 2},
              "stroke": {"value": "rgba(75,200,150,0.8)"}
            },
            "update": {
              "interpolate": {"value": "step-after"}, 
              "fillOpacity": {"value": 1}
            },
            "hover": {
              "fillOpacity": {"value": 0.5}
            }
          }
        },
         {
              "type": "symbol",
              "from": {"data": "series"},
              "encode": {
                "enter": {
                  "x": {"scale": "x", "field": "x"},
                  "y": {"scale": "y", "field": "y"},        
                  "stroke": {"value": "rgba(75,200,150,0.8)"},
                  "fill": {"value": "rgba(75,200,150,0.8)"}
                },
                "update": {        
                  "x": {"scale": "x", "field": "x"},
                  "y": {"scale": "y", "field": "y"},
                  "size": {"value": "20"},
                  "shape": {"value": "square"},
                  "opacity": {"value": 1},
                  "strokeWidth": {"value":2},
                  "fillOpacity": {"value": 1},
                  "strokeOpacity": {"value": 1}
                },
                "hover": {
                  "opacity": {"value": 0.5},
                }
              }
            },{
            "type": "text",
            "interactive": false,
            "encode":{
            "x": {"scale": "x", "value": 100},
            "y": {"scale": "x", "value": 100},
            "fill": {"value": "black"},
            "fontSize": {"value": 20},
            "align": {"value": "right"}
            },
            "update": {
              "text": {"signal": "mytip"}
            }
            }
    ]
    },
               
  ]
};


--------------------------------------------------------------------------------------------

  var vegaJSONSpec = {
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "width": 778,
  "height": 389,
  "padding": 50,
  "signals": [
    {
      "name": "interpolate",
      "value": "step",
      },
    {
      "name": "tooltip",
      "value": {},
      "on": [
        {"events": "symbol:mouseover", "update": "datum"},
        {"events": "symbol:mouseout",  "update": "{}"}
      ]
    },
    {
      "name": "mytip",
      "value": "Test",
      "update": "tooltip ? 'hovered' : 'Test'"
    }     
      ],
      
  "data": [
    {
      "name": "table",
      "format": {"type":"json", "parse":{"x":"number"}},
      "values": [
    
{x: 1495053866000 , y:5},
{x: 1495053871000 , y:8},
{x: 1495053876000 , y:10},
{x: 1495053881000 , y:13},
{x: 1495053886000 , y:15},
{x: 1495053891000 , y:17},
{x: 1495053896000 , y:20},
{x: 1495053901000 , y:22},
{x: 1495053906000 , y:25},
{x: 1495053911000 , y:27},
{x: 1495053916000 , y:30},
{x: 1495053921000 , y:32},
{x: 1495053926000 , y:35},
{x: 1495053931000 , y:37},
{x: 1495053936000 , y:40},
{x: 1495053941000 , y:42},
{x: 1495053946000 , y:45},
{x: 1495053951000 , y:47},
{x: 1495053956000 , y:50},
{x: 1495057561000 , y:49},
{x: 1495057566000 , y:46},
{x: 1495057571000 , y:45},
{x: 1495057576000 , y:42},
{x: 1495057581000 , y:39},
{x: 1495057586000 , y:38},
{x: 1495057591000 , y:33},
{x: 1495057596000 , y:29},
{x: 1495057601000 , y:27},
{x: 1495057606000 , y:24},
{x: 1495057611000 , y:19},
{x: 1495057616000 , y:19},
{x: 1495057621000 , y:18},
{x: 1495057626000 , y:16},
{x: 1495057631000 , y:14},
{x: 1495057636000 , y:12},
{x: 1495057641000 , y:9},
{x: 1495057646000 , y:6},
{x: 1495057651000 , y:5},
{x: 1495057656000 , y:3},
{x: 1495057661000 , y:2},
{x: 1495057666000 , y:1},
{x: 1495057671000 , y:1},
{x: 1495057676000 , y:0},
                  
      ]
    }
  ],

  "scales": [
    {
      "name": "x",
      "type": "time",
      "range": "width",
      "domain": [1495053866000, 1495057766000],
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "nice": true,
      "zero": true,
      "domain": {"data": "table", "field": "y"}      
    },
    {
      "name": "color",
      "type": "ordinal",
      "range": "category",
      "domain": {"data": "table", "field": "c"}
    },
    {    
        "name": "scale_convert",
        "type": "ordinal",
        "round": false,
        "domain": [1495053866000, 1495054166000, 1495054466000, 1495054766000, 1495055066000,1495055366000,1495055666000,1495055966000,1495056266000,1495056566000,1495056866000,1495057166000,1495057466000,1495057766000],
        "range": ["00:00", "00:05", "00:10", "00:15", "00:20","00:25","00:30","00:35","00:40","00:45","00:50","00:55","01:00","01:05"]
   }    
  ],

  "axes": [
    {
      "orient": "bottom", 
      "scale": "x",
      "format": "02d", 
      "tickCount": 2,      
      "grid": true,
      "titlePadding": 10,
      "title": "Time (hh:mm)",  
      "values": [1495053866000, 1495054166000, 1495054466000, 1495054766000, 1495055066000,1495055366000,1495055666000,1495055966000,1495056266000,1495056566000,1495056866000,1495057166000,1495057466000, 1495057766000],
      "encode": {
        "title": {
          "update": {
            "fontSize": {"value": 14},
            "font": {"value": "Verdana, Arial"},
          }
        },
        "labels": {
          "update": {
            "fontSize": {"value": 10},
            "font": {"value": "Verdana, Arial"},
            "text": {"scale": "scale_convert", "field": "value"},
            }
          }          
      }
    },
    {"orient": "left", "scale": "y", "grid": true, "title": "Count","titlePadding": 10,
      "encode": {
        "title": {
          "update": {
            "fontSize": {"value": 14},
            "font": {"value": "Verdana, Arial"},
          
          }
        },
        "labels": {
          "update": {
            "fontSize": {"value": 10},
            "font": {"value": "Verdana, Arial"},
            }
          }          
      }
  }
  ],  
  "marks": [  
    {
      "type": "group",
      "from": {
        "facet": {
          "name": "series",
          "data": "table",
          "groupby": "c"
        }
      },
      "marks": [              
        {
          "type": "line",
          "from": {"data": "series"},
          "encode": {
            "enter": {
              "x": {"scale": "x", "field": "x"},
              "y": {"scale": "y", "field": "y"},
              "strokeWidth": {"value": 2},
              "stroke": {"value": "rgba(75,200,150,0.8)"}
            },
            "update": {
              "interpolate": {"value": "step-after"}, 
              "fillOpacity": {"value": 1}
            },
            "hover": {
              "fillOpacity": {"value": 0.5}
            }
          }
        },
         {
              "type": "symbol",
              "from": {"data": "series"},
              "encode": {
                "enter": {
                  "x": {"scale": "x", "field": "x"},
                  "y": {"scale": "y", "field": "y"},        
                  "stroke": {"value": "rgba(75,200,150,0.8)"},
                  "fill": {"value": "rgba(75,200,150,0.8)"}
                },
                "update": {        
                  "x": {"scale": "x", "field": "x"},
                  "y": {"scale": "y", "field": "y"},
                  "size": {"value": "20"},
                  "shape": {"value": "square"},
                  "opacity": {"value": 1},
                  "strokeWidth": {"value":2},
                  "fillOpacity": {"value": 1},
                  "strokeOpacity": {"value": 1}
                },
                "hover": {
                  "opacity": {"value": 0.5},
                }
              }
            }   
      ]
    }
  ]
};

**Comment by tnavda** [comment]
Jeff,

I figured out the problem.  The issue is my first time is 20:44, and since it isn't a multiple of the tick intervals (5min in this case) it jacks the axis label positions up.  If I change the first time to be 20:40 (1495053600000) then it accurately puts the label 00:00 at the start of the X-Axis.  

The other alternative you suggested was instead of putting the actual time in the data table putting the relative time already computed.  This solves the problem above, and by storing the original time in the data structure I can use vega-tooltip to show the actual value when highlighted.  I should have seen this before as correct but when I first tried I forgot to remove the date subtraction and my labels were all wrong.  But it looks good now.  Next piece will be getting zoom working on it, so hopefully labels still give me what I'm looking for.

Thanks for the help!
